src/main/java/com/google/starfish/servlets/HandleNotesServlet.java

```diff
@@ -1,58 +1,128 @@
-import java.net.MalformedURLException;
-import java.net.URL;
 import java.util.List;
 import java.util.Map;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
+import javax.servlet.ServletException;
 import javax.servlet.http.HttpServletResponse;
 import javax.servlet.http.HttpSession;  
 import javax.servlet.http.Cookie;  
-
+import com.google.starfish.services.LabelService;
+import com.google.starfish.services.MiscNoteLabelService;
+import javax.sql.DataSource;
+import java.sql.Connection;  
+import java.sql.SQLException;  
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.Date;
+import java.util.Calendar;
+
+/** Servlet to handle note uploading and adding associated labels to the database */
 @WebServlet("/handle-notes")
 public class HandleNotesServlet extends HttpServlet {
 
   private BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
   private final String COOKIE_NAME = "SFCookie";
+  private LabelService labelService = new LabelService();
+  private MiscNoteLabelService miscNoteLabelService = new MiscNoteLabelService();
   
   @Override
-  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
     if(!validateUser(request)) {
       response.setStatus(HttpServletResponse.SC_FORBIDDEN);
       return;
     }
 
     HttpSession activeSession = request.getSession(false);
     String userId = (String) activeSession.getAttribute("user_id");
-
     String blobKey = getUploadedFileBlobKey(request, "file");
-    if(blobKey == null) {
+    String sourceUrl = request.getParameter("sourceUrl");
+
+    // There must be some source, either a blob or url, for this note to be posted
+    if (blobKey == null && sourceUrl == null) {
       response.setStatus(HttpServletResponse.SC_NOT_ACCEPTABLE);
       return;
     }
-    String title = request.getParameter("title").toLowerCase();
+
+    String title = request.getParameter("title");
     String school = request.getParameter("school").toLowerCase();
     String course = request.getParameter("course").toLowerCase();
-    String[] miscLabels = request.getParameter("miscLabels").split(",");
-
-    // TODO: Put labels into database
+    String[] miscLabels = request.getParameterValues("miscLabels");
 
-    // TODO: Put note into database
+    DataSource pool = (DataSource) request.getServletContext().getAttribute("my-pool");
 
-    response.setStatus(HttpServletResponse.SC_OK);
+    // Insert all labels before inserting note due to foreign key constraint
+    labelService.insertSchoolLabel(pool, school);
+    labelService.insertCourseLabel(pool, course);
+    for (String miscLabel : miscLabels) {
+      labelService.insertMiscLabel(pool, miscLabel.toLowerCase());
+    }
+    
```

Comment:
Even though it doesn't seem like much, I might extract this out into a helper to keep this servlet method as condensed as possible since there's already quite a lot going on below that can't really be condensed

Reply:
You're definitely right here, there's a lot going on and the more that can be abstracted out the more readable and maintainable this code becomes!